### ISSUE TITLE
Can't seem to get Ampache DLNA to work

### ISSUE THREAD

**Issue by drder66** [ampache/ampache#1547]
Apologies for noob unfamilarity, but I'm having trouble getting the DLNA interface to work (independently of the DAAP/RSP interface in my other ticket)

I'm trying to follow the instructions:

> You must enable UPnP Backend in System settings.

Ok, done

> You must also enable Enable url rewriting in Streaming settings.

I think this is done, but what user level to use ? I've tried ticking the "all" box, and "user" and "guest" and "admin" all to no avail.

> If your installation is within a subfolder '/ampache' you must take care for your Rewrite Rules within '[ampachepath]/play/.htaccess'.

It is in /ampache, so I modified .htaccess thus:
IfModule mod_rewrite.c
    RewriteEngine On
    RewriteCond %{REQUEST_FILENAME} !-d
    RewriteCond %{REQUEST_FILENAME} !-s
    RewriteRule ^art/([^/]+)/([^/]+)/([0-9]+)/thumb([0-9]*)\.([a-z]+)$ /ampache/image.php?object_type=$2&object_id=$3&auth=$1&thumb=$4&name=art.jpg [L]
    RewriteRule ^([^/]+)/([^/]+)/([^/]+)/([^/]+)(/.*)?$ /ampache/play/$5?$1=$2&$3=$4 [N,QSA]
    RewriteRule ^([^/]+)/([^/]+)(/.*)?$ /ampache/play/$3?$1=$2 [N,QSA]
    RewriteRule ^(/[^/]+|[^/]+/|/?)$ /ampache/play/index.php [L,QSA]
/IfModule

> PHP Socket mod must be enabled and PHP XMLReader extension available

Now you're losing me. Do you mean PHP socket modification, or PHP socket module ? How would I know which and whether they were enabled ? And from what I can find out the XML reader has been default enabled in the binary releases for some time. 

> You should broadcast the server periodically calling http://localhost/ampache

/upnp/?btnSend=Send+SSDP+broadcast or php bin/broadcast.inc with a cron (every 1-2 minutes recommended).
This seems pretty ugly as it's filling either my apache logs or cron logs or both. For now I've been using the /ampache/upnp/?btnSendAuto=Send+SSDP+broadcast+every+second link to do this, which has duly been filling apache logs up with 

192.168.10.66 - - [26/May/2017:23:52:04 +0100] "GET /ampache/upnp/?btnSendAuto=Send+SSDP+broadcast+every+second HTTP/1.1" 200 778 "-" "Mozilla/5.0 (X11; Ubuntu; Linux i686; rv:53.0) Gecko/20100101 Firefox/53.0"
192.168.10.66 - - [26/May/2017:23:52:06 +0100] "GET /ampache/upnp/images/upnp.jpg HTTP/1.1" 304 181 "http://cxxxxxxxx.linkpc.net/ampache/upnp/?btnSendAuto=Send+SSDP+broadcast+every+second" "Mozilla/5.0 (X11; Ubuntu; Linux i686; rv:53.0) Gecko/20100101 Firefox/53.0"

> You need to set the http_host option within config/ampache.cfg.php to broadcast via broadcast.inc.

This is not clear. I've set http_host to cxxxxxxx.linkpc.net, but it it needs to be set to something else this needs clearer explanation. What is broadcasting via broadcast.inc ? ping cxxxxxxx.linkpc.net resolves to 192.168.10.2 which is the server's address.

I am using a non-standard TCP port for the web server by default, or https (it's not clear how to handle this other than leave the port commented out to be auto filled in). Not sure if that's relevant to uPNP anyway, as it's not clear how uPNP requests are handled in the first place.

I did add a route for a broadcast domain
239.0.0.0       *               255.0.0.0       U     0      0        0 br0
But this made no difference

Any help appreciated. 

# Environment

 * Ampache version: 3.8.2
 * Web server + version: Apache/2.4.10 (Debian) mod_fcgid/2.3.9 OpenSSL/1.0
 * Server operating system: Debian Jesse (ARM)
 * Client operating system: Several different clients tried

 * Client type:
   - UPnP Tried ubuntu client; can see other upnp devices on the network, but not Ampache. Viera TV can also see these devices but not ampache

# Settings

;#<?php exit(); ?>##
;#########################################################
; General Config                                         #
;#########################################################

; This value is used to detect quickly
; if this config file is up to date
; this is compared against a value hard-coded
; into the init script
config_version = 33

;#########################################################
; Path Vars                                              #
;#########################################################

; The public http host of your server.
; If not set, retrieved automatically from client request.
; This setting is required for WebSocket server
; DEFAULT: ""
http_host = "cxxxxxxxxx.linkpc.net"

; The public http port of your server.
; If not set, retrieved automatically from client request.
; DEFAULT:
;http_port = 80

; The public path to your ampache install
; Do not put a trailing / on this path
; For example if your site is located at http://localhost
; than you do not need to enter anything for the web_path
; if it is located at http://localhost/music you need to
; set web_path to /music
; DEFAULT: ""
web_path = "/ampache"

; The local http url of your server.
; If not set, retrieved automatically from server information.
; DEFAULT: ""
;local_web_path = "http://localhost/ampache"

;#########################################################
; Database                                               #
;#########################################################

; Hostname of your database
; For socket authentication, set the path to socket file (e.g. /var/run/mysqld/mysqld.sock)
; DEFAULT: localhost
database_hostname = "localhost"

; Port to use when connecting to your database
; DEFAULT: none
database_port = ""

; Name of your ampache database
; DEFAULT: ampache
database_name = "ampache"

; Username for your ampache database
; DEFAULT: ""
database_username = "ampache"

; Password for your ampache database, this can not be blank
; this is a 'forced' security precaution, the default value
; will not work (except if using socket authentication)
; DEFAULT: ""
database_password = "xxxxxxxxxx"

(rest of settings per default install)

# Description

# Logs

2017-05-25 22:13:04 [admin] (session) -> f015jir9bq19qhudbot4r6gsd6 has been extended to Thu, 25 May 2017 23:13:04 +0100 extension length 3600
2017-05-25 22:13:04 [admin] (ajax.server.php) -> Called for page: {index}
2017-05-25 22:13:07 [ampache] (Plugins) -> Lastfm.plugin.php is not a directory.
2017-05-25 22:13:07 [ampache] (Plugins) -> Librefm.plugin.php is not a directory.
2017-05-25 22:13:07 [ampache] (Plugins) -> MusicBrainz.plugin.php is not a directory.
2017-05-25 22:13:11 [ampache] (Plugins) -> Lastfm.plugin.php is not a directory.
2017-05-25 22:13:11 [ampache] (Plugins) -> Librefm.plugin.php is not a directory.
2017-05-25 22:13:11 [ampache] (Plugins) -> MusicBrainz.plugin.php is not a directory.

There's nothing in the webserver logs other than occasionally:
www.cxxxxxxxxxxxxx.linkpc.net:3689 ::1 - - [26/May/2017:22:43:54 +0100] "OPTIONS * HTTP/1.0" 200 126 "-" "Apache/2.4.10 (Debian) mod_fcgid/2.3.9 OpenSSL/1.0.1t (internal dummy connection)"

PS: what is all this "plugin is not a directory" stuff anyway ? I don't have any plugins loaded.

**Comment by drder66** [comment]
Update:
Added the following to my apache default site:

ProxyPass /ws ws://127.0.0.1:8100 retry=0
ProxyPassReverse /ws ws://127.0.0.1:8100 retry=0
ProxyRequests off
ProxyTimeout 15

This allows a Panasonic TV to access the server via uPNP. However, a Roberts Stream 202 just hangs loading media information. GUPnP is able to load tracks, but not play them.

Nothing interesting in the logs. It would be good to be able to turn on detailed logging somehow to find out what is going on.

**Comment by wagnered** [comment]
@drder66,

Sorry nobody has replied to you on this.  I'm not familiar with the DLNA feature, but I'll try to give  you as much help as I can.

> Do you mean PHP socket modification, or PHP socket module ? 
It's the php modules and at the command line, type php -m  to print out a list of loaded modules.  'socket' and 'XMLReader' should be in the list.

The 'all' box overrides the individual levels and gives permissions to all levels.

I assume you are referring to the Ampache/DLNA server?  I'll look into it. We are short on developers and they are very busy with higher priority commitments.

**Comment by drder66** [comment]
xmlreader and socket**s** (plural) are there when php -m is run. I assume this is ok.

Any hints as to how to get any logging for DLNA operations ? I have debug=true and debug_level = 5 but am not seeing anything very interesting recorded

**Comment by wagnered** [comment]
Your http_host indicates that linkpc.net is your domain.  Does this mean that you are behind a router? If so, Do you have port forwarding set?

I suggest that http_host is set to default and attempt to get it working from the local network first.

I'm using lighttpd as my web server, so I can't help with apache stuff.

>You should broadcast the server periodically calling http://localhost/ampache...

The full url is needed. you are getting a 304 (not modified) error which means, I think, that the rewrite rules are being ignored.

**Comment by wagnered** [comment]
rereading previus replies, I just noticed:

>This allows a Panasonic TV to access the server via uPNP.

And files can be played?

Try using VLC. It is supposed to contain it's own controller.  Open the playlist and undock it if it is.  The left pane allows searching for UPNP sources. 

>You need to set the http_host option within config/ampache.cfg.php to broadcast via broadcast.inc.

This is only necessary if not using the wiki's previous suggestion.

**Comment by wagnered** [comment]
@drder66,

I got it!

Uncomment the line `;websocket_address = "ws://localhost:8100"` in ampache.cfg.php.

I'll edit the instructions in the wiki to include this requirement.

**Comment by wagnered** [comment]
OOPs, this only enabled listing of Ampache server music in gUPnP AV control control point. They don't seem to play.

**Comment by wagnered** [comment]
You indicate you are using Apache.  Do you have mod-rewrite enabled? 
Rewriting is definitely necessary.

73 From  AA1AD
Ernie D
Federal Way, WA | CN87ug

**Comment by drder66** [comment]
Thanks Wagnered; as for mod-rewite:
> a2enmod rewrite
Module rewrite already enabled

I uncommented websocket_address, but no difference. I tried to go back to what was working, but alas that no longer works (I can't see Ampache from the DLNA client in the TV)

I'm puzzled by port 8100 in the websocket, as I see nothing DLNA-related for that TCP Port number (1900 seems to be associated with DLNA).

It would really help if there was some way to get some logging going for this interface. All I can see in Ampache logs is some session information when I go in normally through the web interface.

**Comment by wagnered** [comment]
Don't forget to run 'php bin/broadcast` with cron or
`http://localhost/ampache/upnp/?btnSend=Send+SSDP+broadcast` from the
browser. The Ampache DLNA server must "Notify" it's service for any
control point to notice it.

Well, I don't know much about Apache, but I did see that `AllowOverride
FileInfo` or `|AllowOverride All` needs to be added to Apache's config
file as found in
(https://askubuntu.com/questions/48362/how-to-enable-mod-rewrite-in-apache)[https://askubuntu.com/questions/48362/how-to-enable-mod-rewrite-in-apache].|

|I've been doing a lot of reading on uPnP (DLNA, as I understand is a
subset of uPnP) and there is still a lot I have yet to learn. But the
requested file is accessed through http.  I set up kodi which does have
logging. At least until the first error, then it no longer logs if it
continues to not find the access the file: That's http 404. Kodi uses
curl to request the selected file.  So I copied the URL  and manually
attempted access.  I finally had to end up formatting the request as a
URL with a query string which means to me that rewrite rules are needed.|

|I also used wireshark to see what was being sent through the interface.
|

|Because the error is from the web server, not Ampache, logging needs to
be enabled in the server.  As I think I wrote earlier, I use Lighttpd
which can log access and php can log errors.
|

|I'm still working on the problem.|

|
|

73 From  AA1AD
Ernie D
Federal Way, WA | CN87ug

**Comment by drder66** [comment]
Bit more progress: using iptraf on the router/server I identified that the broadcasts weren't happening. Then I remembered that one needs to set up a broadcast route on the server:

`route add 239.255.255.250 br0`  (or to the appropriate interface(s)) and that last time I'd done this I'd not made the change permanent. 

This has allowed the TV (Panasonic 40DS500 series) to see Ampache **and** play music again, which is good. There's now plenty of stuff going on the Ampache logs once Ampache is selected on the TV.

The Stream 202 player and Pansonic BWT735 PVR on the same wlan both refuse to see Ampache. Both see other DLNA devices on the network (e.g. minidlna, running on my laptop). Like the TV, they both sees these devices **immediately**, unlike Ampache, which has to wait up to a minute to be discovered when the broadcast goes off. I can't help feeling that Ampache (or at least my setup) is missing some part of the discovery mechanism - like receiving some kind of broadcast when another DLNA client is looking for servers to connect to (is that what ws:8100 is about ???). Short of quietening down the whole network and doing extensive wiresharking, and without logging I can't think what a good next step would be.

As for VLC, I can't get that to see anything, not even the DLNA streams from miniDLNA that everything else can see.

**Comment by drder66** [comment]
One more discovery; logging some packets on the router (I'm only using iptraf at the moment so don't have the packet contents) when I ask the TV to look for DLNA devices, there's a whole load of non-IP traffic going on, some of which is broadcast (to address 0xFFFFFFFFFFFF).
I don't have much knowledge of PHP, but I've grepped over the sources and I don't find any references to handling anything to do with layer 2 packets at all. Would you agree ?

**Comment by drder66** [comment]
Oh and I should say I'm not trying to use DNLA outside the local LAN

**Comment by wagnered** [comment]
@drder66,

I just installed apache and updated php to 7.0 from the Fedora repository.

I refreshed Kobi's playlist from the Ampache server and started playing a file.  It is now playing.

guPnp sees the Ampache server, but I don't know how to set a renderer to play the music.

Totem now sees the Ampache server, but crashes when I dig down into the music list. It never saw the Ampache server before.

For VLC, it is necessary to start Ampache's sspd broadcasts before starting VLC.  It then will see the Ampache server. Double clicking the entry will cause VLC to start listing the files.  They play on VLC.

I posted my httpd.conf at [https://gist.github.com/wagnered/f275d6527606c9ffde2c1c18b48170de](https://gist.github.com/wagnered/f275d6527606c9ffde2c1c18b48170de) for comparison.

**Comment by drder66** [comment]
Fundamentally, I think something is missing.

I see no process on my server listening on port 1900: this is the uPNP discovery port that the client pings a broadcast message to in order to find out what servers are out there to connect to.

If the client happens to hang around long enough, it may choose to notice Ampache sending out an SSDP discovery at some later time. Or it may just decide there's no-one there because Ampache didn't respond it its discovery broadcast.

There's nothing listening on port 8100 either. I'm not sure that the websocket_address is relevant. What do you think ?

**Comment by wagnered** [comment]
As I stated previously after I installed Apache, the whole system
started seeing the server and playing the songs/videos: guPnP, kobi,
vlc, Totem all have built in controllers and did not need stand-alone
controllers such as mimidlna, or the upnp universal control point to
function.

So I know it works with Apache.  I couldn't figure out how to set up the
rewrite rules for lighttpd; That's why I installed Apache (httpd on
Fedora). The rewrite rules are already set up in play/.htaccess.

I posted my Apache config file at 
(https://gist.github.com/wagnered/f275d6527606c9ffde2c1c18b48170de)[https://gist.github.com/wagnered/f275d6527606c9ffde2c1c18b48170de]
for you to check and see if yours is set up correctly.

So I know it works.  I suggest you set up /bin/broadcast.inc on cron to
broadcast every minute. or
http://localhost/ampache/upnp/?btnSend=Send+SSDP+broadcast and have it
broadcast every second while you are debugging your set up.  Like I
wrote earlier, VLC and possibly others must see the broadcast while
starting.


73 From  AA1AD
Ernie D
Federal Way, WA | CN87ug

**Comment by drder66** [comment]
I have a cron job which is calling wget to wake the ssdp web page up; I know it's working because without that the TV (the only device which can see Ampache) can't see Ampache.

Other uPNP servers appear immediately when I start the TV or PVR or Stream 202. Ampache shows up on the TV as soon as the cron job fires.

To me this means that the clients are sending out their SSDP's, getting no response from Ampache and then assuming that there's nothing there.

There's an interesting page at ampache/upnp/find.php which produces the following response:
```

Array
(
    [0] => stdClass Object
        (
            [CACHE-CONTROL] => max-age=1800
            [DATE] => Thu, 22 Jun 2017 21:17:28 GMT
            [EXT] => 
            [LOCATION] => http://192.168.10.143:8008/ssdp/device-desc.xml
            [OPT] => "http://schemas.upnp.org/upnp/1/0/"; ns=01
            [01-NLS] => b36b3e68-1dd1-11b2-bb59-c26696b07cde
            [SERVER] => Linux/3.8.13+, UPnP/1.0, Portable SDK for UPnP devices/1.6.18
            [X-USER-AGENT] => redsonic
            [ST] => upnp:rootdevice
            [USN] => uuid:7fa1824b-ae7b-5b42-fe73-e31614d5842f::upnp:rootdevice
            [BOOTID.UPNP.ORG] => 501
            [CONFIGID.UPNP.ORG] => 2
        )

    [1] => stdClass Object
        (
            [CACHE-CONTROL] => max-age=1800
            [EXT] => 
            [LOCATION] => http://192.168.10.123:55000/dms/ddd.xml
            [SERVER] => FreeBSD/8.0 UPnP/1.0 Panasonic-MIL-DLNA-SV/1.0
            [ST] => upnp:rootdevice
            [USN] => uuid:4D454930-0000-1000-8001-A81374A786E0::upnp:rootdevice
        )

    [2] => stdClass Object
        (
            [CACHE-CONTROL] => max-age=1800
            [EXT] => 
            [LOCATION] => http://192.168.10.123:55000/dmr/ddd.xml
            [SERVER] => FreeBSD/8.0 UPnP/1.0 Panasonic-MIL-DLNA-SV/1.0
            [ST] => upnp:rootdevice
            [USN] => uuid:4D454930-0100-1000-8001-A81374A786E0::upnp:rootdevice
        )

    [3] => stdClass Object
        (
            [CACHE-CONTROL] => max-age=1800
            [EXT] => 
            [LOCATION] => http://192.168.10.123:55000/nrc/ddd.xml
            [SERVER] => FreeBSD/8.0 UPnP/1.0 Panasonic-MIL-DLNA-SV/1.0
            [ST] => upnp:rootdevice
            [USN] => uuid:4D454930-0200-1000-8001-A81374A786E0::upnp:rootdevice
        )

    [4] => stdClass Object
        (
            [CACHE-CONTROL] => max-age=1800
            [DATE] => Thu, 22 Jun 2017 21:17:28 GMT
            [ST] => upnp:rootdevice
            [USN] => uuid:4d696e69-444c-164e-9d41-000ae42cff1e::upnp:rootdevice
            [EXT] => 
            [SERVER] => Ubuntu DLNADOC/1.50 UPnP/1.0 MiniDLNA/1.1.5
            [LOCATION] => http://192.168.10.63:8200/rootDesc.xml
            [CONTENT-LENGTH] => 0
        )

    [5] => stdClass Object
        (
            [CACHE-CONTROL] => max-age=1800
            [DATE] => Thu, 22 Jun 2017 21:17:28 GMT
            [EXT] => 
            [LOCATION] => http://192.168.10.128:60607/8CC121B21B40/Server1/ddd
            [SERVER] => Linux/4.0 UPnP/1.0 Panasonic-UPnP-MW/1.0
            [ST] => upnp:rootdevice
            [USN] => uuid:4D454930-0300-1000-8000-8CC121B21B40::upnp:rootdevice
        )

    [6] => stdClass Object
        (
            [CACHE-CONTROL] => max-age=1800
            [DATE] => Thu, 22 Jun 2017 21:17:28 GMT
            [EXT] => 
            [LOCATION] => http://192.168.10.128:60606/8CC121B21B40/Server0/ddd
            [SERVER] => Linux/4.0 UPnP/1.0 Panasonic-UPnP-MW/1.0
            [ST] => upnp:rootdevice
            [USN] => uuid:4D454930-0100-1000-8000-8CC121B21B40::upnp:rootdevice
        )

)

This seems to be what the clients are doing: sending out a broadcast to port 1900 and seeing what's out there. I don't see Ampache providing this service.
```

**Comment by wagnered** [comment]
To quote the wiki: 
> ...or php bin/broadcast.inc with a cron (every 1-2 minutes recommended).

I use `* * * * *       env DISPLAY=:0 /usr/bin/php -f /home/ernie/htdocs-mars/ampache/bin/broadcast.inc ` in my crontab file.  This gives a broadcast every minute.

If you actually paste the url "http://localhost/ampache/upnp/?btnSend=Send+SSDP+broadcast" into a web page, you will see two buttons to push: One to broadcast every second, the other is a "one-shot".  wget cannot push any of those buttons. I'm not sure why your TV is seeing the server without Ampache sending out a NOTIFY broadcast.

If you don't see:
    CMDOUT (Starting broadcasts...)
    CMDOUT (UPnP broadcast... Done.)
You will know your crontab command is not working

**Comment by wagnered** [comment]
I was referring to cron.log.

>If you don't see:
CMDOUT (Starting broadcasts...)
CMDOUT (UPnP broadcast... Done.)
you will know your crontab command is not working

**Comment by drder66** [comment]
The cron job /is/ working. If I stop it, the TV doesn't see Ampache.

Unfortunately with the cron job working and the TV able to play music, the other two devices just don't see Ampache at all, ever. The do see minidlna and each other (the TV/PVR are servers and clients) and any other uPNP server I choose to fire up, like windows media server.

What do you see when you do 
`lsof -i :1900`  ?
IMHO there should be a listner on this port, able to respond to SSDP probes from the clients when they try to enumerate servers in their domain.

**Comment by wagnered** [comment]
I forgot to ask what version of Ampache are you using?  You might try
out 3.8.3-develop if you are not using it: A few uPnP issues were fixed.
That's what I've been using.

Ampache doesn't have a "listener". It doesn't respond to the M-SEARCH
message. That's why it needs to continuously broadcast the NOTIFY message.

If the problem isn't with the Ampache version or your Apache config
file, I don't know what else to suggest.  I have it working here, so I
know there is no problem with Ampache's (partial) DLNA service.

73 From  AA1AD
Ernie D
Federal Way, WA | CN87ug

**Comment by wagnered** [comment]
BTW I will look into the possibility of adding the ssdp listener to the Ampache/Laravel version.

**Comment by drder66** [comment]
I'm currently using 3.8.2 (was up there in the original post). I'll try and get the latest loaded soon.

Getting an ssdp listener in sounds like a great idea. If you need any help testing, I'm happy to do what I can. I was getting ready to get some wireshark traces from the devices that won't connect anyway.

BTW is there anything in node-ssdp that might help that ?

**Comment by drder66** [comment]
Just confirmed that the server is receiving M-SEARCH requests from the Stream202 radio using 
`https://gist.github.com/provegard/1435555`
But it doesn't seem interested in the SSDP announcements that Ampache subsequently sends each minute

**Comment by wagnered** [comment]
I posted my ampache.cfg.php for comparison at
|https://gist.github.com/anonymous/789a2c2398ba11e134652301172f696d|.

I fired up my  Window 7 VM and had the same problem. Apps could see the
minidlna on the Linux server, but not the Ampache uPnP service.


73 From  AA1AD
Ernie D
Federal Way, WA | CN87ug

**Comment by alejandroliu** [comment]
This issue is from 2017 (2 years old).  Are people still having problems with this?  Is this working?
I myself managed to get DLNA to work with VLC.
Allthough I did have to fix a number of issues in the code so that it works for me now.

**Comment by lachlan-00** [comment]
No idea, the DLNA code looks pretty incomplete so it wouldn't surprise me if people were still having issues.

I haven't used it but i do have enough DLNA capable devices that i should be able to work on it.

**Comment by f1xmAn** [comment]
> This issue is from 2017 (2 years old). Are people still having problems with this? Is this working?
> I myself managed to get DLNA to work with VLC.
> Allthough I did have to fix a number of issues in the code so that it works for me now.

Hi, I'm still having issues. I've done everything described in the Wiki, however, I didn't manage VLC to see Ampache DLNA server. How did you managed to configure it? I'm using Apache.

**Comment by drder66** [comment]
I'm still having issues with this. If I set up the default route and force SSDP pings I can get VLC to see the server, but there are no playlists etc.

**Comment by drder66** [comment]
A proper description of everything necessary to get DLNA working (and debugged when some step doesn't work, because, perhaps ampache is in a different place on the web server) would be immensely useful if there is anyone who has it working who could spare the time to write it up

**Comment by lachlan-00** [comment]
I don't have it active but I'm sure I'd had it running before.

We've been documenting a lot this year so it should be added to the wiki when I get started again.

**Comment by shikasta-net** [comment]
I'm just now having exactly all these problems.  I've added complexity of running Ampache inside docker and have now migrated my miniDLNA service to use `macvlan` instead of `host` network, which was _the_ blocker for using Ampache's DLNA instead.  That all works nicely but I can't see Ampache's DLNA, although I can access it via it's (new) LAN IP, so the networking part is solid but VLC is not picking up the broadcasts.  The only setting I've changed is the UPNP inside the admin > system menu because I can't find enough documentation to be confident fiddling with anything else.

I'd be happy to try @alejandroliu suggestions if I can build a docker image with your modifications.

**Comment by drder66** [comment]
Any update on this? I'm on to version 4.1 but this doesn't seem to have improved anything

**Comment by lachlan-00** [comment]
no work has been done on this. I need to work out how it's all done and document it as it's not a feature i've used previously

**Comment by drder66** [comment]
I'm wondering if it's worth considering hacking gerbera to use the ampache database schema. The upnp code there works well and gerbera's mysql schema is considerably simpler. 
Thinking to compose the query results for gerbera from ampache queries and disable gerbera scanning if it finds the ampache schema
Thoughts?

**Comment by drder66** [comment]
I did some more work on this. 

The reason most DLNA clients don't discover Ampache reliably is that Ampache does not listen for or respond to M-SEARCH messages. It relies on the client noticing NOTIFY messages. Some do, some don't.

I wrote an SSDP listener in Python (I did this because writing stuff to make Apache handle UDP seems tough). I don't make great claims for programming skills, but this solves most of the problems with getting Ampache noticed. This has been tested on:
Panasonic TV, Roberts Stream-202, VLC, Windows 10 Media Player, eezUPNP, gupnp-universal-cp and a few others. It works with everything except the Stream-202, which is is unable to resolve a qualified domain name (it works if I make the http_host be the local IP address, but then I can't use Ampache outside the house. Working on this separately).  I can make this available if people are interested. 

So much for the good news. The only client to get anywhere near working properly is the Panasonic TV. This shows Video/Music and allows browsing the catalog. Not every song in every album shows correctly. Those that do can't play, because the TV sends back the play HTTP request in slash format but Ampache wants to see something like ampache/play/index.php?ssid=xxxxx&player=api& etc. I hacked on .htaccess in /ampache/play 
<IfModule mod_rewrite.c>
    RewriteEngine On
    RewriteCond %{REQUEST_FILENAME} !-d
    RewriteCond %{REQUEST_FILENAME} !-s
    RewriteRule ^art/([^/]+)/([^/]+)/([0-9]+)/thumb([0-9]*)\.([a-z]+)$ ../image.php?object_type=$2&object_id=$3&auth=$1&thumb=$4&name=art.jpg [L]
    RewriteRule "ssid/([^/]+)/([^/]+)/([^/]+)/([^/]+)/([^/]+)/([^/]+)/([^/]+)/[^/]+/[^/]+/([^/]+)/([^/]+)" index.php?ssid=$1&$2=$3&$4=$5&$6=$7&player=api&$8=$9 [N,QSA]
</IfModule>

But this really wants to have mod_rewrite change more than 10 parameters which requires more thought than I have time for. I haven't figured out how/why the client ends up asking in this way. But for now, the above hack on .htaccess does allow the TV to play ok.

The rest is a bit of a disaster. 
VLC sits in a hard loop asking for the root items (Music/Video), being sent them and asking again indefinitely. But VLC won't play list or things off my Pansonic DVR either, where Media Player will. 
For Ampache, Windows10 Media player doesn't get to list the root elements, nor does eezUPNP. I'm guessing there's something non-compliant that VLC (and the others) are choking on but the TV is happy to ignore. 
Access to the DLNA spec would be one thing; getting Ampache checked by a DLNA test suite would be better still. Without either, I'm stuck guessing what might be wrong with the code in control-reply.php

**Comment by drder66** [comment]
I've done quite a bit of work on this now and have upnp working across a much wider range of clients. 
I'd like to discuss folding in these changes - is there somewhere I can have an offline discussion with the maintainers?

**Comment by lachlan-00** [comment]
Sure, you can email if you want and i can reply back from my own address (security [at] ampache [dot] dev)

**Comment by drder66** [comment]
I did write last week - not sure if you received.

**Comment by lachlan-00** [comment]
Got it! I've just been out of town for the last week so I'm working through the backlog

**Comment by ravenclaw900** [comment]
I'm kind of having the same issue.  I can get Ampache to show on VLC but it won't undock or play any songs.

**Comment by drder66** [comment]
Hi

What version of Ampache are you using?

drdr6

**Comment by ravenclaw900** [comment]
4.4.1

**Comment by drder66** [comment]
Ok, that should have fixes in it that allow Ampache to work

Do you just see the first line of Ampache: All our music and cannot open 
Ampache line?

**Comment by ravenclaw900** [comment]
Yes, it shows up as 'Ampache :: For the Love of Music', but when I double-click on it, it just moves on to the next server.

**Comment by drder66** [comment]
VLC version & OS ? (Mine is 3.0.9.2, Ubuntu)

Is there anything showing up in Ampache's log file? (debug = "true", 
debug_level = "5" in ampache.cfg.php)

**Comment by drder66** [comment]
Hi,

I may have the answer to your problem. It seems that in the grand scheme 
of things, strict php compliance has been turned on and this has broken 
a bit of the upnp api. If you can open ./lib/class/upnp_api.php and go 
to line 273:

change

     $result = XMLReader::XML($prmRequest);
to

     $result = $reader->XML($prmRequest);

Let me know if this fixes the problem (it does for me when I load a 
clean instance of 4.4.1)

Thanks

drdr6

**Comment by ravenclaw900** [comment]
Getting there!  Ampache itself will undock (and so will Videos), but Music won't do anything.

**Comment by drder66** [comment]
Are there any errors being reported in the log file ?

If not, can you go into the config file, uncomment the debug section 
(setting it to true) and send what gets dumped to the log file when you 
try to open Music ?

Ta

drdr6

**Comment by ravenclaw900** [comment]
Those SQL errors at the bottom look like problems.
```log
2021-03-30T18:08:07-05:00 [ampache] (Upnp_Api) -> Envelope 1 1  
2021-03-30T18:08:07-05:00 [ampache] (Upnp_Api) -> #text 14 1  
2021-03-30T18:08:07-05:00 [ampache] (Upnp_Api) -> Body 1 1  
2021-03-30T18:08:07-05:00 [ampache] (Upnp_Api) -> Browse 1 1  
2021-03-30T18:08:07-05:00 [ampache] (Upnp_Api) -> #text 14 1  
2021-03-30T18:08:07-05:00 [ampache] (Upnp_Api) -> ObjectID 1 1  
2021-03-30T18:08:07-05:00 [ampache] (Upnp_Api) -> ObjectID 15 1  
2021-03-30T18:08:07-05:00 [ampache] (Upnp_Api) -> #text 14 1  
2021-03-30T18:08:07-05:00 [ampache] (Upnp_Api) -> BrowseFlag 1 1  
2021-03-30T18:08:07-05:00 [ampache] (Upnp_Api) -> BrowseFlag 15 1  
2021-03-30T18:08:07-05:00 [ampache] (Upnp_Api) -> #text 14 1  
2021-03-30T18:08:07-05:00 [ampache] (Upnp_Api) -> Filter 1 1  
2021-03-30T18:08:07-05:00 [ampache] (Upnp_Api) -> Filter 15 1  
2021-03-30T18:08:07-05:00 [ampache] (Upnp_Api) -> #text 14 1  
2021-03-30T18:08:07-05:00 [ampache] (Upnp_Api) -> StartingIndex 1 1  
2021-03-30T18:08:07-05:00 [ampache] (Upnp_Api) -> StartingIndex 15 1  
2021-03-30T18:08:07-05:00 [ampache] (Upnp_Api) -> #text 14 1  
2021-03-30T18:08:07-05:00 [ampache] (Upnp_Api) -> RequestedCount 1 1  
2021-03-30T18:08:07-05:00 [ampache] (Upnp_Api) -> RequestedCount 15 1  
2021-03-30T18:08:07-05:00 [ampache] (Upnp_Api) -> #text 14 1  
2021-03-30T18:08:07-05:00 [ampache] (Upnp_Api) -> SortCriteria 1 1  
2021-03-30T18:08:07-05:00 [ampache] (Upnp_Api) -> #text 14 1  
2021-03-30T18:08:07-05:00 [ampache] (Upnp_Api) -> Browse 15 1  
2021-03-30T18:08:07-05:00 [ampache] (Upnp_Api) -> #text 14 1  
2021-03-30T18:08:07-05:00 [ampache] (Upnp_Api) -> Body 15 1  
2021-03-30T18:08:07-05:00 [ampache] (Upnp_Api) -> #text 14 1  
2021-03-30T18:08:07-05:00 [ampache] (Upnp_Api) -> Envelope 15 1  
2021-03-30T18:08:07-05:00 [ampache] (control-reply) -> Request: <s:Envelope xmlns:s="http://schemas.xmlsoap.org/soap/envelope/" s:encodingStyle="http://schemas. mlsoap.org/soap/encoding/">
2021-03-30T18:08:07-05:00 [ampache] (control-reply) -> <s:Body><u:Browse xmlns:u "urn:schemas-upnp-org:service:ContentDirectory:1">
2021-03-30T18:08:07-05:00 [ampache] (control-reply) -> <ObjectID>amp://music</Ob ectID>
2021-03-30T18:08:07-05:00 [ampache] (control-reply) -> <BrowseFlag>BrowseDirectC ildren</BrowseFlag>
 021-03-30T18:08:07-05:00 [ampache] (control-reply) -> <Filter>*</Filter>
2021-03-30T18:08:07-05:00 [ampache] (control-reply) -> <StartingIndex>0</Startin Index>
2021-03-30T18:08:07-05:00 [ampache] (control-reply) -> <RequestedCount>5000</Req estedCount>
2021-03-30T18:08:07-05:00 [ampache] (control-reply) -> <SortCriteria></SortCrite ia>
 021-03-30T18:08:07-05:00 [ampache] (control-reply) -> </u:Browse>
 021-03-30T18:08:07-05:00 [ampache] (control-reply) -> </s:Body>
 021-03-30T18:08:07-05:00 [ampache] (control-reply) -> </s:Envelope>
 021-03-30T18:08:07-05:00 [ampache] (control-reply) -> 
2021-03-30T18:08:07-05:00 [ampache] (control-reply) ->  
2021-03-30T18:08:07-05:00 [ampache] (control-reply) -> Action: browse with filter [*] 
2021-03-30T18:08:07-05:00 [ampache] (control-reply) -> ObjectID: amp://music 
2021-03-30T18:08:07-05:00 [ampache] (control-reply) -> Listrequest  
2021-03-30T18:08:07-05:00 [ampache] (Upnp_Api) -> MusicChilds: [] [][0] [5000] 
2021-03-30T18:08:07-05:00 [ampache] (Upnp_Api) -> MusicChilds4: [] 
2021-03-30T18:08:07-05:00 [ampache] (Dba) -> Error_query SQL: SELECT COUNT(`id`) FROM `label` 
2021-03-30T18:08:07-05:00 [ampache] (Dba) -> Error_query MSG: ["42S02",1146,"Table 'ampache.label' doesn't exist"] 
2021-03-30T18:08:07-05:00 [ampache] (Dba) -> Error_query SQL: SELECT COUNT(`id`) FROM `label` 
2021-03-30T18:08:07-05:00 [ampache] (Dba) -> Error_query MSG: ["42S02",1146,"Table 'ampache.label' doesn't exist"] 
2021-03-30T18:08:07-05:00 [ampache] (Dba) -> Error_query SQL: SELECT COUNT(`id`) FROM `label` 
2021-03-30T18:08:07-05:00 [ampache] (Dba) -> Error_query MSG: ["42S02",1146,"Table 'ampache.label' doesn't exist"] 
2021-03-30T18:08:07-05:00 [ampache] (Dba) -> Error_query SQL: SELECT COUNT(`id`) FROM `label` 
2021-03-30T18:08:07-05:00 [ampache] (Dba) -> Error_query MSG: ["42S02",1146,"Table 'ampache.label' doesn't exist"] 
```

**Comment by wagnered** [comment]
@ravenclaw900 

I'm guessing you have a damaged database.  The "label" table is created upon initial setup..

**Comment by wagnered** [comment]
if you don't want to reinstall Ampache, phpmyadmin or `mysql` command line can be used;

copy 
```
DROP TABLE IF EXISTS `label`;
CREATE TABLE IF NOT EXISTS `label` (
  `id` int(11) UNSIGNED NOT NULL AUTO_INCREMENT,
  `name` varchar(80) CHARACTER SET utf8 COLLATE utf8_general_ci DEFAULT NULL,
  `category` varchar(40) CHARACTER SET utf8 COLLATE utf8_general_ci DEFAULT NULL,
  `summary` text CHARACTER SET utf8 COLLATE utf8_general_ci DEFAULT NULL,
  `address` varchar(256) CHARACTER SET utf8 COLLATE utf8_general_ci DEFAULT NULL,
  `email` varchar(128) CHARACTER SET utf8 COLLATE utf8_general_ci DEFAULT NULL,
  `website` varchar(256) CHARACTER SET utf8 COLLATE utf8_general_ci DEFAULT NULL,
  `user` int(11) UNSIGNED DEFAULT NULL,
  `creation_date` int(11) UNSIGNED DEFAULT NULL,
  PRIMARY KEY (`id`)
) ENGINE=MyISAM DEFAULT CHARSET=utf8 COLLATE=utf8_unicode_ci;
```

into a file and import it with phpmyadmin or execute `mysql -u USER -p DATABASE_NAME < PATH/TO/FILE.sql` in the cli.

The above code is taken from ampache.sql.

**Comment by ravenclaw900** [comment]
Yes, that fixed it.  Thank you!

**Comment by shikasta-net** [comment]
Some progress made today with regard to getting it working inside docker.
With the container of 5.4 from the official repo, all that's needed is to enable the UPnP module and rewrite (which seem to be by default).
Attach to the network described by this compose snippet
`networks:
    broadcast:                                                                           
    driver: macvlan                                                                    
    driver_opts:                                                                       
      parent: eth0 <- the host's network adapter                                                                   
    ipam:                                                                              
      config:                                                                          
        - subnet: 192.168.0.0/24   <- the LAN subnet                                                    
          ip-range: 192.168.0.4/30    <- constrain IP assignemnt                                                 
          gateway: 192.168.0.1          <- LAN gateway                                               
          aux-address:                                                                 
            SERV001: 192.168.0.2  <- IPs already in use `
and attach the container with this compose snippet:
`ampache:
   ...
   networks:                                                        
      broadcast:                                                                
        ipv4_address: 192.168.0.6 `

**Comment by shikasta-net** [comment]
I've solved the network visibility problem:
docker compose attaches networks in alphabetical order, even if you use the `priority` property as above.  *The broadcast network must be the first in the container's interface list*; when you run `docker inspect` the order of the networks is the order they were added. With my macvlan broadcast network at the top of the list, the service works, but now I have the pre-existing problem that VLC can't browse the content even with the 1second refresh (which has been discussed in this thread).

**Comment by cnic023** [comment]
Can't seem to make any headway getting upnp backend to work. Renderers show up on BubbleUPNP and gssdp-discover (a SamsungTV and a Pioneer receiver). But neither Bubble or VLC can 'see' ampache. I have broadcast working every minute in cron. I think I've gone through all the checklists but no luck. I'm using ampache 5.5.2-release (no docker) and PHP 8.0. Would make my year to get this working. Ampache is GREAT!

**Comment by cnic023** [comment]
In case it helps there follow the results of 'gssdp-discover -i [name of ethernet device on the ampache server]
192.168.0.21 is the Samsung series 6 TV
192.168.0.52 is the Pioneer receiver
http://:/upnp/MediaServerServiceDesc.php would appear to be Ampache?? I don't see what else it could be.
Running gssdp-discover on another server on the same subnet produces the same result so it appears to me as a noob that the ampache broadcast is working. Perhaps it needs to happen more frequently than every minute?
****************************************************************

Scanning for all resources
Showing "available" messages
resource available
  USN:      uuid:d6b98f4d-e5e9-47c0-9509-81835785878c::upnp:rootdevice
  Location: http://192.168.0.21:7678/nservice/
resource available
  USN:      uuid:d6b98f4d-e5e9-47c0-9509-81835785878c
  Location: http://192.168.0.21:7678/nservice/
resource available
  USN:      uuid:d6b98f4d-e5e9-47c0-9509-81835785878c::urn:dial-multiscreen-org:device:dialreceiver:1
  Location: http://192.168.0.21:7678/nservice/
resource available
  USN:      uuid:d6b98f4d-e5e9-47c0-9509-81835785878c::urn:dial-multiscreen-org:service:dial:1
  Location: http://192.168.0.21:7678/nservice/
resource available
  USN:      uuid:6e71e96f-85d5-4c67-bfd3-11060e096716::upnp:rootdevice
  Location: http://192.168.0.21:9197/dmr
resource available
  USN:      uuid:6e71e96f-85d5-4c67-bfd3-11060e096716
  Location: http://192.168.0.21:9197/dmr
resource available
  USN:      uuid:6e71e96f-85d5-4c67-bfd3-11060e096716::urn:schemas-upnp-org:device:MediaRenderer:1
  Location: http://192.168.0.21:9197/dmr
resource available
  USN:      uuid:6e71e96f-85d5-4c67-bfd3-11060e096716::urn:schemas-upnp-org:service:RenderingControl:1
  Location: http://192.168.0.21:9197/dmr
resource available
  USN:      uuid:6e71e96f-85d5-4c67-bfd3-11060e096716::urn:schemas-upnp-org:service:ConnectionManager:1
  Location: http://192.168.0.21:9197/dmr
resource available
  USN:      uuid:6e71e96f-85d5-4c67-bfd3-11060e096716::urn:schemas-upnp-org:service:AVTransport:1
  Location: http://192.168.0.21:9197/dmr
resource available
  USN:      uuid:b0f2e8a0-00a0-e8BB-b000-0009b0f2e8a0::urn:schemas-upnp-org:service:RenderingControl:2
  Location: http://192.168.0.52:8888/upnp_descriptor_0
resource available
  USN:      uuid:SSTV-KS20-0000000000000001370895::upnp:rootdevice
  Location: http://192.168.0.21:9080
resource available
  USN:      uuid:b0f2e8a0-00a0-e8BB-b000-0009b0f2e8a0::urn:schemas-upnp-org:service:AVTransport:2
  Location: http://192.168.0.52:8888/upnp_descriptor_0
resource available
  USN:      uuid:SSTV-KS20-0000000000000001370895
  Location: http://192.168.0.21:9080
resource available
  USN:      uuid:SSTV-KS20-0000000000000001370895::urn:schemas-upnp-org:device:mdxdevice:1
  Location: http://192.168.0.21:9080
resource available
  USN:      uuid:SSTV-KS20-0000000000000001370895::urn:mdx-netflix-com:service:target:3
  Location: http://192.168.0.21:9080
resource available
  USN:      uuid:b0f2e8a0-00a0-e8BB-b000-0009b0f2e8a0
  Location: http://192.168.0.52:8888/upnp_descriptor_0
resource available
  USN:      uuid:b0f2e8a0-00a0-e8BB-b000-0009b0f2e8a0::upnp:rootdevice
  Location: http://192.168.0.52:8888/upnp_descriptor_0
resource available
  USN:      uuid:b0f2e8a0-00a0-e8BB-b000-0009b0f2e8a0::urn:schemas-upnp-org:service:ConnectionManager:2
  Location: http://192.168.0.52:8888/upnp_descriptor_0
resource available
  USN:      uuid:b0f2e8a0-00a0-e8BB-b000-0009b0f2e8a0::urn:schemas-upnp-org:device:MediaRenderer:2
  Location: http://192.168.0.52:8888/upnp_descriptor_0
resource available
  USN:      uuid:953b053e-eb4f-c42b-c603-2f472c7ad238::upnp:rootdevice
  Location: http://:/upnp/MediaServerServiceDesc.php
resource available
  USN:      uuid:953b053e-eb4f-c42b-c603-2f472c7ad238
  Location: http://:/upnp/MediaServerServiceDesc.php
resource available
  USN:      uuid:953b053e-eb4f-c42b-c603-2f472c7ad238::urn:schemas-upnp-org:device:MediaServer:1
  Location: http://:/upnp/MediaServerServiceDesc.php
resource available
  USN:      uuid:953b053e-eb4f-c42b-c603-2f472c7ad238::urn:schemas-upnp-org:service:ConnectionManager:1
  Location: http://:/upnp/MediaServerServiceDesc.php
resource available
  USN:      uuid:953b053e-eb4f-c42b-c603-2f472c7ad238::urn:schemas-upnp-org:service:ContentDirectory:1
  Location: http://:/upnp/MediaServerServiceDesc.php

**Comment by cnic023** [comment]
Following is a recurring extract from my ampache log.
2022-10-07T09:28:51+13:00 [cnic023] (ajax.server) -> Called for page: {index}
2022-10-07T09:28:51+13:00 [cnic023] (Ampache\Module\System\Session) -> Writing to ip2kjdhkbtdga4toeqcvhs7536 with expiration 1665091731
2022-10-07T09:28:54+13:00 [ampache] (Ampache\Module\System\Dba) -> Database connection...
2022-10-07T09:28:54+13:00 [ampache] (Ampache\Module\System\Dba) -> Database connection...
2022-10-07T09:28:54+13:00 [cnic023] (Ampache\Module\System\Session) -> gp8i3ji46n19fm4gshi7vldev3 has been extended to Fri, 07 Oct 2022 10:28:54 +1300 extension length 3600

**Comment by cnic023** [comment]
Sorry for my abysmal ignorance but I have been assuming Ampache is the controller and my TV and Receiver are the renderers. However, I can't see how, in Ampache, I get to choose which renderer the data goes to. Or ... am I missing something here such as a phone app or similar that I need?

**Comment by drder66** [comment]
I did some work on this a while back and Ampache is lacking primarily in documentation for uPnp

The SSDP broadcast every minute I found   inadequate for many devices so I wrote some ptp that can be run as a daemon to handle the device inquiries as well as the broadcast. Look for the SSDP service code.

Unfortunately I don't have time to look at this / write much more for a couple of weeks owing to work commitments. 







On 6 October 2022 22:06:18 BST, cnic023 ***@***.***> wrote:
>Sorry for my abysmal ignorance but I have been assuming Ampache is the controller and my TV and Receiver are the renderers. However, I can't see how, in Ampache, I get to choose which renderer the data goes to. Or ... am I missing something here such as a phone app or similar that I need?
>
>-- 
>Reply to this email directly or view it on GitHub:
>https://github.com/ampache/ampache/issues/1547#issuecomment-1270698289
>You are receiving this because you were mentioned.
>
>Message ID: ***@***.***>
-- 
Sent from my Android device with K-9 Mail. Please excuse my brevity.

**Comment by cnic023** [comment]
Thanks for that. Intuitively it seems to be all or part of my problem

Get Outlook for Android<https://aka.ms/AAb9ysg>
________________________________

**Comment by cnic023** [comment]
Thanks once again for your help on this. Unfortunately, while I can do some Perl scripting, I'm totally ignorant about PHP. I've found what looks like a service script in 'public/upnp' but I don't know how to integrate it into Ampache. For all I know it may already be integrated. In any event, running the script didn't seem to achieve anything. Another thing that puzzles me is whether, once the service is operational, it needs to have the cron broadcasts continue in conjunction?
I have tried to get the upnp backend running for a very long time now because I can see how much it could enhance my present Ampache setup. But I don't think I have the skills or a proper overview of what is happening (or is supposed to happen).
Anyway, I appreciate your advice. It is good to know that it is possible to get it running. Previous threads threw some doubt on this.
Best wishes
Chris

**Comment by drder66** [comment]
Lachlan (if you're monitoring this thread still), or any other 
maintainer familiar with PlayAction.php

I'm trying to help user cnic023 get uPnP working in Ampache 5.x. 
Unfortunately there seem to have been breakages since I checked in.

So far only one or two small things until I came to trying to play a 
file, whereupon the code falls completely flat. The code in 
PlayAction.php no longer functions at all owing to the changed semantics 
of how arguments are being passed in from Apache.

I had placed some code starting with "$slashcount = 
substr_count($_SERVER['QUERY_STRING'], '/');" to replace the server 
query string to break a URL based string into a set of arguments that 
are then picked up by the code starting "$action       = 
(string)filter_input(INPUT_GET, 'action', FILTER_SANITIZE_SPECIAL_CHARS);"

I think somewhere else you've managed to change the meaning of INPUT_GET 
from an alias to "_SERVER" to something involving 
Nyholm\Psr7Server\ServerRequestCreatorInterface - but I can't find the 
definition of filter_input. I can find _filter_input, but that doesn't 
seem to relate to the arguments being passed to filter_input and I can't 
find the definition of INPUT_GET anywhere.

Given that in uPnP mode, the input to PlayAction.php is different, and 
that the string ought to be bounds checked before touching it, I'd 
appreciate your thoughts on how to modify PlayAction to cope with both 
input formats (unless you can tell me that PlayAction isn't used for 
anything else like DAAP or some other API that I'm not testing). I did 
it the way I did it previously in order to minimize the new code, but it 
may be that whatever INPUT_GET is now pointing to is no longer 
writeable, so a parallel set of decodes for all the components may be 
necessary for uPnP

Thanks
drdr66

**Comment by drder66** [comment]
I've returned to this to try to get things working, starting from the 6.2.1 code base. A lot more has changed in the PlayAction.php code since last time, and attempting to use this for DLNA causes breakages in numerous places. 

The fundamental problem appears to be that PlayAction.php is more than ever being expected to be called by a valid user, and with DLNA there is no user (uid=0, and DLNA/upnp doesn't have a concept of users or access rights). So it's not surprising that the test at line 344 

if (make_bool($user->disabled))  

fails because $user->disabled isn't initialized. 

And the at line 400 we fail again because of not having a user; if we force this to pass because we do have a session ID 
we fail again around line 525 with "Stream control failed for user %s" - because again we don't have a user.

I'd like to ask what the design philosophy here is: is Ampache now locked so firmly into the concept of being driven by a "user" (with rights to a catalog etc) that a user-less API like DLNA is not going to work (or be too much effort to get to work), and bypassing all the user code checks/lookups etc. also presenting a potential security hazard? 
Should I soldier on putting bypasses for active session ID against this code; or should we consider perhaps having to have  a "dlna_user" ID in the system when using upnp so that upnp can pass that user ID out and when PlayAction gets called to play a song it ends up being asked to do so with the non-zero user ID of this pseudo-user?

**Comment by richard-lott** [comment]
> I've returned to this to try to get things working, starting from the 6.2.1 code base. A lot more has changed in the PlayAction.php code since last time, and attempting to use this for DLNA causes breakages in numerous places.

I've been trying to get this working myself just recently and I've run into these same issues you have. I'd love to help out but my PHP is basically none existent.

> I'd like to ask what the design philosophy here is: is Ampache now locked so firmly into the concept of being driven by a "user" (with rights to a catalog etc) that a user-less API like DLNA is not going to work (or be too much effort to get to work), and bypassing all the user code checks/lookups etc. also presenting a potential security hazard? Should I soldier on putting bypasses for active session ID against this code; or should we consider perhaps having to have a "dlna_user" ID in the system when using upnp so that upnp can pass that user ID out and when PlayAction gets called to play a song it ends up being asked to do so with the non-zero user ID of this pseudo-user?

Jellyfin allows you to select a user to be the default when using DLNA. Maybe you could do something similar to that, then people have the option to select a regular account or to create a DLNA specific user?

**Comment by lachlan-00** [comment]
we had a contributor fix up a lot of upnp issues a while ago so it should be resolved
